async_server.c

```diff
@@ -0,0 +1,381 @@
+#include "async_server.h"
+
+#include "brute.h"
+#include "log.h"
+#include "multi.h"
+#include "server_common.h"
+
+#include <arpa/inet.h>
+#include <errno.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <pthread.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#ifdef __FreeBSD__
+#include <sys/socket.h>
+#include <sys/types.h>
+#endif
+
+static acl_context_t *
+acl_context_init (serv_context_t *global_ctx)
+{
+  acl_context_t *ctx = calloc (1, sizeof (*ctx));
+  if (!ctx)
+    {
+      error ("Could not allocate memory for client context");
+      return (NULL);
+    }
+  ctx->context = global_ctx;
+
+  if (queue_init (&ctx->registry_idx, sizeof (int)) != QS_SUCCESS)
+    {
+      error ("Could not initialize registry indices queue");
+      goto cleanup;
+    }
+  for (int i = 0; i < QUEUE_SIZE; ++i)
+    if (queue_push (&ctx->registry_idx, &i) != QS_SUCCESS)
+      {
+        error ("Could not push index to registry indices queue");
+        goto cleanup;
+      }
+
+  ctx->ref_count = 1;
+  if (pthread_mutex_init (&ctx->mutex, NULL) != 0)
+    {
+      error ("Could not initialize mutex");
+      goto cleanup;
+    }
+
+  return ctx;
+
+cleanup:
+  if (queue_cancel (&ctx->registry_idx) != QS_SUCCESS)
+    error ("Could not cancel registry indices queue");
+
+  if (queue_destroy (&ctx->registry_idx) != QS_SUCCESS)
+    error ("Could not destroy registry indices queue");
+
+  free (ctx);
+
+  return (NULL);
+}
+
+static void
+acl_context_destroy (acl_context_t *ctx)
+{
+  trace ("Destroying client context");
+
+  if (queue_cancel (&ctx->registry_idx) != QS_SUCCESS)
+    error ("Could not cancel registry indices queue");
+
+  trace ("Cancelled registry indices queue");
+
+  if (queue_destroy (&ctx->registry_idx) != QS_SUCCESS)
+    error ("Could not destroy registry indices queue");
+
+  trace ("Destroyed registry indices queue");
+
+  if (pthread_mutex_destroy (&ctx->mutex) != 0)
+    error ("Could not destroy mutex");
+
+  close_client (ctx->socket_fd);
+  free (ctx);
+}
+
+static status_t
+return_tasks (acl_context_t *ctx)
+{
+  mt_context_t *mt_ctx = &ctx->context->context;
+  status_t status = S_SUCCESS;
+
+  int i;
+  for (i = 0; i < QUEUE_SIZE; ++i)
+    {
+      if (ctx->registry_used[i])
+        {
+          if (queue_push_back (&mt_ctx->queue, &ctx->registry[i])
+              != QS_SUCCESS)
+            {
+              status = S_FAILURE;
+              break;
+            }
+          ctx->registry_used[i] = false;
+        }
+    }
+
+  return (status);
+}
+
+static void
+sender_receiver_cleanup (void *arg)
+{
+  acl_context_t *ctx = arg;
+
+  bool is_last;
+  if (pthread_mutex_lock (&ctx->mutex) != 0)
+    return;
+  pthread_cleanup_push (cleanup_mutex_unlock, &ctx->mutex);
```

Comment:
I'm still hesitant about return_tasks under lock.

As an alternative aproach, we could store receiver/sender threads ips in client context and explicitly thread_cancel both. Then return_tasks could be moved to if (is_last) branch after lock.

Reply:
I added a TODO mark to try to do this with a fresh head after I do some tasks other than async client-server, so can we squash & merge now?